contracts/src/GaspToken.sol

```diff
@@ -1,30 +1,126 @@
+contract GaspToken is Context, Ownable, ERC20, IGaspToken {
+    uint256 private constant _TOTAL_SUPPLY = 1_000_000_000 * 10 ** 18;
+    string private constant _NAME = "GASP";
+    string private constant _SYMBOL = "GASP";
+
+    mapping(address => bool) public override whitelist;
+    address public override uniswapPool;
+    bool public override allowTransfers;
+
+    modifier isWhitelisted(address[2] memory addresses, bytes4 selector) {
+        if (!allowTransfers) {
+            if (!whitelist[addresses[0]] && !whitelist[addresses[1]] || addresses[1] == uniswapPool) {
+                revert OperationForbidden(selector);
+            }
+        }
+        _;
     }
 
-    function mint(address account, uint256 amount) public {
-      require((owner == msg.sender), "Only one who deployed contract can mint tokens");
-      _mint(account, amount);
+    constructor(address l1Council_) Ownable() ERC20(_NAME, _SYMBOL) {
+        if (l1Council_ == address(0)) {
+            revert ZeroL1Council();
+        }
+
+        _transferOwnership(l1Council_);
+        _mint(l1Council_, _TOTAL_SUPPLY);
+
+        whitelist[l1Council_] = true;
     }
-}
 
+    function setUniswapPool(address uniswapPool_) external override onlyOwner {
+        if (uniswapPool_ == address(0)) {
+            revert ZeroUniswapPool();
+        }
 
-contract Gasp is MintableERC20 {
-    constructor() MintableERC20("GaspV2", "GASPV2") {}
-}
+        uniswapPool = uniswapPool_;
+        whitelist[uniswapPool_] = true;
+        emit UniswapPoolSet(uniswapPool_);
+    }
 
-contract DevToken is MintableERC20 {
-    constructor() MintableERC20("GaspDev", "GASPDEV") {}
-}
+    function setAllowTransfers(bool allowTransfers_) external override onlyOwner {
+        if (allowTransfers) {
+            revert TransfersAlreadyAllowed();
+        }
+
+        allowTransfers = allowTransfers_;
+        emit AllowTransfersSet(allowTransfers_);
+    }
+
+    function addToWhitelist(address account) external override onlyOwner {
+        if (account == address(0)) {
+            revert ZeroWhitelistAccount();
+        }
+        if (whitelist[account]) {
+            revert AccountAlreadyWhitelisted(account);
+        }
+
+        whitelist[account] = true;
+        emit AddedToWhitelist(account);
+    }
 
-contract ArbToken is MintableERC20 {
-    constructor() MintableERC20("CrossChain Token", "CCXT") {}
-}
\ No newline at end of file
+    function removeFromWhitelist(address account) external override onlyOwner {
+        if (account == address(0)) {
+            revert ZeroWhitelistAccount();
+        }
+        if (!whitelist[account]) {
+            revert AccountNotWhitelisted(account);
+        }
+
+        delete whitelist[account];
+        emit RemovedFromWhitelist(account);
+    }
+
+    function approve(address spender, uint256 amount)
+        public
+        override(ERC20, IERC20)
+        isWhitelisted([_msgSender(), spender], IERC20.approve.selector)
```

Comment:
All the isWhitelisted calls seem to be in violation of 
"Transfer of tokens, tokens can leave only from addresses included in the "FROM whitelist""

Reply:
This changed during Serges direct discussions with Gleb and Stano. In the end, they decided to make only one whitelist for simpler logic and fewer edge cases, which I actually appreciate.

Reply:
See my first comprehensive comment above.

Reply:
Including recipients into a whitelist was never discussed or approved. A whitelisted address should be able to send tokens to non-whitelisted addresses. An address, who received the token, is not obliged to be whitelisted.

Violating this requirement will end up in a situation when we are unable to send tokens to airdrop participants, investors, anyone without including them into the whitelist- thus allowing them to transfer tokens to other whitelisted addresses.